Item stat percentage info and identification reorder

kristofbolyai opened the pull request:
TODO:

- [x] Implement API route parsing and add all necessary classes
- [x] Add `ItemsReceivedEvent`
- [x] Add `ItemTooltipRenderEvent`
- [x] Add `ItemStatInfoFeature` feature
  - [x] Identification stats (percentage, min, max, reroll chances)
     - [x] Fix issues related and test
  - [x] Item stat reordering
  - [x] Item stat spacing
- [x] Fix `ItemGuessFeature` and `PlayerGhostTransparencyFeature`

Incompleteusern committed:
Make feature constructor nicer to newInstance

kristofbolyai committed:
Parse everything related to `/cache/get/itemList`

kristofbolyai committed:
Spotless.

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/IdentificationOrderer.java:
```diff
@@ -0,0 +1,103 @@
+    public IdentificationOrderer(
+            Map<String, Integer> idOrders,
+            ArrayList<String> groupRanges,
+            ArrayList<String> inveverted) {}
```
misspelt

Incompleteusern commented on common/src/main/java/com/wynntils/wc/objects/items/IdentificationContainer.java:
```diff
@@ -0,0 +1,235 @@
+import java.util.Map;
+import org.apache.commons.lang3.math.Fraction;
+
+public class IdentificationContainer {
```
This code has always seem convoluted and hard to read to me

Incompleteusern reviewed (commented):
Some things, nice progress

kristofbolyai committed:
Fix typo

Incompleteusern committed:
rewrite IdentificationContainer to be cleaner and work better

I hope so at least

Incompleteusern commented on common/src/main/java/com/wynntils/wc/objects/items/IdentificationContainer.java:
```diff
@@ -0,0 +1,190 @@
+    public static String getAsLongName(String shortName) {
+        if (shortName.startsWith("raw")) {
+            shortName = shortName.substring(3);
+            shortName = Character.toLowerCase(shortName.charAt(0)) + shortName.substring(1);
```
Maybe make a `StringUtils` method for this too?

kristofbolyai replied:
I don't currently see the need, but if this code will be needed elsewhere, we can move it.

Incompleteusern commented:
This is evidently a future thing but now since `DynamicTag` exists it can be used for rainbow and defective names which is beneficial. I don't think it should be used for the actual lore though.

kristofbolyai committed:
Fix typo

kristofbolyai committed:
Added `ItemsReceivedEvent`

kristofbolyai committed:
Added `ItemTooltipRenderEvent`

Incompleteusern committed:
Add remaining item received hooks

Incompleteusern committed:
update features to use new event when applicable

Incompleteusern committed:
fix optional usage in ComponentUtils

kristofbolyai committed:
Added the first iteration of stat overlays (percentage overlay)

kristofbolyai commented:
Forgot to add a NBT tag so we don't reparse the item lore every time. Noting it so we don't forget.

Incompleteusern commented on common/src/main/java/com/wynntils/features/ItemStatPercentageInfoFeature.java:
```diff
@@ -0,0 +1,141 @@
+
+            Matcher statusMatcher = ITEM_STATUS_PATTERN.matcher(unformattedLoreLine);
+
+            if (statusMatcher.matches() && !endOfStatuses) {
```
Minor thing but stop applying status matcher after endOfStatuses is reached

Incompleteusern commented on common/src/main/java/com/wynntils/features/ItemStatPercentageInfoFeature.java:
```diff
@@ -0,0 +1,141 @@
+                        profile.getLongNameStatusMap().get(statusName);
+
+                if (idContainer.hasConstantValue()) {
+                    newLore.add(StringTag.valueOf(ItemUtils.toLoreForm(newLoreLine)));
```
Another small thing but replace `StringTag.valueOf(ItemUtils.toLoreForm(newLoreLine))` with `lore.get(i)`

Incompleteusern reviewed (commented):
2 things

kristofbolyai committed:
Optimizations

Incompleteusern committed:
color lerping

Incompleteusern committed:
resolve error

Incompleteusern commented:
personally I want to replace the Keyboard changing lore with formatters

Incompleteusern committed:
implement 0% and 100% items

kristofbolyai committed:
Rename methods

kristofbolyai committed:
Added initial support for format variables

kristofbolyai committed:
Rewrote `ItemStatInfoFeature.replaceLore` to fix some systematic issues

Fixed all the issues with unrecognised ids, and also we know always store the old item info in NBT, so we don't get lost in the changes.

Incompleteusern committed:
push fixes and changes

Incompleteusern committed:
spotless and comment

Incompleteusern committed:
generify

Incompleteusern committed:
fixes

Incompleteusern committed:
spotless

magicus commented on common/src/main/java/com/wynntils/features/ItemStatInfoFeature.java:
```diff
@@ -28,11 +29,28 @@
 
+    // TODO: Replace these with configs
+    // Possible variables: %percentage% %minmax% %reidchance%
+    private static final String MAIN_FORMAT_STRING = "%percentage%";
```
If we're re-inventing the info strings, maybe chose a wiser format this time? Using asymmetric delimiters help parsing, both for human readers and for the code. In other words, `{percentage}` is a better format string than `%percentage%`.

kristofbolyai replied:
I agree, let's refactor this.

magicus replied:
I've also been thinking about not really having "info" variables, as much as having "variables". To expand a bit, we should have some kind of standardized registry mapping String names to lambdas returning a value. We could then use this "source of truth" in our code, but we could also allow users to access it through the "info variable" syntax. It's not a big change in the code itself, it's more of a conceptual change. I'm not sure if such a thing should be part of this PR (probably not, better to have small, dedicated PRs) but I'd like to see something like that at some point in time for Artemis.

kristofbolyai replied:
Sounds good, however note that these are "item based" info variables so it wouldn't really be useful here. However cool idea for  normal info variables.

Incompleteusern committed:
add more info variables

Incompleteusern committed:
fix issues with formatter

Incompleteusern committed:
fix gray

kristofbolyai committed:
Fix issues related to id type parsing

kristofbolyai committed:
Fix bugs

kristofbolyai committed:
Fix italic text

kristofbolyai committed:
Spotless

kristofbolyai committed:
Fix spacing issues

kristofbolyai committed:
Changed formatters

kristofbolyai committed:
Spotless

kristofbolyai commented:
At this point, `ItemStatInfoFeature` works nicely ingame. 
![image](https://user-images.githubusercontent.com/49001742/155188010-353151d7-cb46-46e8-9909-4135995d9d04.png)
![image](https://user-images.githubusercontent.com/49001742/155188014-10bf7e1e-9cc9-45dd-b878-e7a00a9124f1.png)

Let's do some optimizations and fixes and let's move on to `ItemStatIndentificationReorderFeature`?

Incompleteusern commented:
@kristofbolyai I think `ItemStatIndentificationReorderFeature` should be merged with `ItemStatInfoFeature` for the sake of the code, also it'd be easier to access the settings that way

Incompleteusern commented:
A future thing to do when configs are done is to make the keybinds customizable and have a arbitrary number of presets

Incompleteusern committed:
increase precision of chance inc and dec

Incompleteusern committed:
rewrite duplicate removal code

Incompleteusern committed:
forgot to increment index

Incompleteusern committed:
use better colors for lerping

Incompleteusern committed:
fix more derps

Incompleteusern committed:
Fix italics

Incompleteusern committed:
dont apply green/red formatting to info var addition

kristofbolyai committed:
Fixed issues related to inverted ids

Incompleteusern committed:
account for inverted stars

Incompleteusern committed:
oops

kristofbolyai committed:
Item lore reordering and grouping

Incompleteusern committed:
use addall

kristofbolyai committed:
Fix `ItemGuessFeature` and `PlayerGhostTransparencyFeature`

kristofbolyai commented:
Also note, players were a ghost even though `PlayerGhostTransparencyFeature` was not working. I believe this is caused by the mixin/event calling. Investigate further later.

EDIT: I actually might be misremembering, could be easily checked with NPCs (Market NPC) though.

Incompleteusern commented:
but why doesn't `ItemsRecievedEvent` work

kristofbolyai commented:
> but why doesn't `ItemsRecievedEvent` work

I believe it either did not fire every inventory update or for some reason it did not always update the lore. Also `ItemToolTipRenderEvent` is more efficient imo, it kinda acts as a "lazy loading" way, while `ItemsReceivedEvent` forces calculations every inventory update.

kristofbolyai committed:
Fix `ItemStatInfoFeature` lore duplication

Incompleteusern committed:
make settings

Incompleteusern committed:
fix ItemGuessesProfile with ItemsReceivedEvent

Adds price info too

Incompleteusern committed:
forgot to add

Incompleteusern committed:
add onDisables for ones missing them

Incompleteusern commented:
New color lerped image
![image](https://user-images.githubusercontent.com/58920010/155565708-1e785dc6-fc6d-4f6c-9afc-b170c727696b.png)
Also this has been done finally too
![image](https://user-images.githubusercontent.com/58920010/155565808-62fcdf00-a9ab-445a-ba50-99a618ad713d.png)

Incompleteusern reviewed (dismissed):
Given that I've worked on this and have tested it, I'm fine with this.

Incompleteusern committed:
fix bug issue

kristofbolyai committed:
Use `groupIdentifications`

kristofbolyai committed:
Color alternative format

kristofbolyai commented:
New color for alternative format: 
![image](https://user-images.githubusercontent.com/49001742/155576081-52ba570e-2963-4a5e-9c12-aa1ad8043b5b.png)

Incompleteusern commented on common/src/main/java/com/wynntils/features/ItemStatInfoFeature.java:
```diff
@@ -0,0 +1,593 @@
+
+            // TODO add escaped chars like old impl for special chars
+            infoVariables.put(
+                    "chance_perfect",
```
Don't add the unicode symbol or color to the info variable itself, make it optional for the user

kristofbolyai replied:
If we don't add it there, it will not use the correct color. I think we will have to worry about this when we add configs,

Incompleteusern replied:
Can't you put a normal color code before it? Pretty sure that overrides

kristofbolyai replied:
![image](https://user-images.githubusercontent.com/49001742/155589005-b1fae50b-fb2a-4231-99df-3480d219dbfb.png)

Code: 
```java
            infoVariables.put(
                    "chance_perfect",
                    new TextComponent(
                                    String.format(
                                            "\u2605%.2f%%", idContainer.getPerfectChance() * 100)));
            infoVariables.put(
                    "chance_increase",
                    new TextComponent(String.format("\u21E7%.1f%%", chances.increase() * 100)));
            infoVariables.put(
                    "chance_decrease",
                    new TextComponent(String.format("\u21E9%.1f%%", chances.decrease() * 100)));
```
and
```java
    private static final String ALTERNATIVE_FORMAT_STRING =
            "{percentage} {chance_perfect}"
                    + " {chance_increase} {chance_decrease}"
                    + ChatFormatting.YELLOW
                    + " ["
                    + "{min}"
                    + ChatFormatting.YELLOW
                    + ", "
                    + "{max}"
                    + ChatFormatting.YELLOW
                    + "]"; // Used when user presses SHIFT on lore.
```

This confirms that we have to apply the style there.

Min/Max works because it applies the style:
```java
            infoVariables.put(
                    "min",
                    new TextComponent(String.valueOf(min))
                            .withStyle(
                                    Style.EMPTY
                                            .withColor(ChatFormatting.YELLOW)
                                            .withItalic(false)));
            infoVariables.put(
                    "max",
                    new TextComponent(String.valueOf(max))
                            .withStyle(
                                    Style.EMPTY
                                            .withColor(ChatFormatting.YELLOW)
                                            .withItalic(false)));
```

Incompleteusern replied:
Makes sense but it is rather unhappy to force the colors

kristofbolyai replied:
I agree, but it should be fine for now. We can get back to this when formatters aren't hardcoded.

kristofbolyai committed:
Fix issues related to market

kristofbolyai committed:
Try to force non-italic text by default in `toLoreStringTag`.

kristofbolyai committed:
Add FeatureInfo to `ItemStatInfoFeature`

Incompleteusern commented on common/src/main/java/com/wynntils/features/ItemStatInfoFeature.java:
```diff
@@ -276,7 +276,27 @@ private void replaceLore(ItemStack itemStack, boolean alternativeForm) {
-            newLore.add(idStart, ItemUtils.toLoreStringTag(new TextComponent("")));
+
+            // Handle Market spacing issue
+            if (lore.size() > 4) {
```
Surely there's a better fix. Why is it broken, since I am not exactly sure.

If there's a double space before the first id, maybe strip it first. Or only remove double spaces after first index.

kristofbolyai committed:
More elegant fix to market issues

kristofbolyai committed:
Some italic fixes related to perfect and defective

Incompleteusern committed:
move time call out of loop

kristofbolyai committed:
Remove FIXMEs

kristofbolyai committed:
Fix italic weirdness with defective

kristofbolyai committed:
Actually fix italic weirdness now

kristofbolyai committed:
Fixed broken obfuscation

Incompleteusern committed:
set base component to not be italicized

kristofbolyai committed:
Fix perfect

kristofbolyai committed:
Merge remote-tracking branch 'origin/item_ids' into item_ids

# Conflicts:
#	common/src/main/java/com/wynntils/features/ItemStatInfoFeature.java

kristofbolyai committed:
Remove useless modulo

Incompleteusern committed:
fix obfuscated and new setting
